File: backend/src/main/java/com/trio/backend/service/hr/EmployeeSkillService.java
package com.trio.backend.service.hr;

import com.trio.backend.dto.hr.CreateEmployeeSkillRequest;
import com.trio.backend.dto.hr.EmployeeSkillResponse;
import com.trio.backend.dto.hr.EmployeeSkillSearchCriteria;
import com.trio.backend.dto.hr.EmployeeSkillStatistics;
import com.trio.backend.dto.hr.UpdateEmployeeSkillRequest;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

public interface EmployeeSkillService {

    EmployeeSkillResponse create(UUID workspaceId, UUID departmentId, UUID employeeId, CreateEmployeeSkillRequest request);

    EmployeeSkillResponse getById(UUID workspaceId, UUID departmentId, UUID employeeId, UUID skillId);

    Page<EmployeeSkillResponse> listByEmployee(UUID workspaceId, UUID departmentId, UUID employeeId, Pageable pageable);

    Page<EmployeeSkillResponse> search(UUID workspaceId, UUID departmentId, EmployeeSkillSearchCriteria criteria, Pageable pageable);

    EmployeeSkillResponse update(UUID workspaceId, UUID departmentId, UUID employeeId, UUID skillId, UpdateEmployeeSkillRequest request);

    void delete(UUID workspaceId, UUID departmentId, UUID employeeId, UUID skillId);

    EmployeeSkillResponse verify(UUID workspaceId, UUID departmentId, UUID employeeId, UUID skillId);

    EmployeeSkillResponse unverify(UUID workspaceId, UUID departmentId, UUID employeeId, UUID skillId);

    EmployeeSkillStatistics getStatistics(UUID workspaceId, UUID departmentId);

    List<EmployeeSkillResponse> getExpiringCertifications(UUID workspaceId, UUID departmentId, int withinDays);
}

File: frontend/project/src/services/employee-skill-service.ts
import { apiClient } from '../lib/api';
import type { PageResponse } from '../types/api';

export interface EmployeeSkillResponse {
  id: string;
  employeeId: string;
  skillName: string;
  category: string;
  proficiencyLevel: string;
  yearsOfExperience?: number;
  lastUsedDate?: string;
  certificationName?: string;
  certificationIssuer?: string;
  certificationDate?: string;
  certificationExpiration?: string;
  verified: boolean;
  verifiedBy?: string;
  verifiedAt?: string;
  notes?: string;
  active: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface CreateEmployeeSkillRequest {
  skillName: string;
  category: string;
  proficiencyLevel: string;
  yearsOfExperience?: number;
  lastUsedDate?: string;
  certificationName?: string;
  certificationIssuer?: string;
  certificationDate?: string;
  certificationExpiration?: string;
  notes?: string;
}

export interface UpdateEmployeeSkillRequest {
  skillName?: string;
  category?: string;
  proficiencyLevel?: string;
  yearsOfExperience?: number;
  lastUsedDate?: string;
  certificationName?: string;
  certificationIssuer?: string;
  certificationDate?: string;
  certificationExpiration?: string;
  notes?: string;
  active?: boolean;
}

export interface EmployeeSkillStatistics {
  totalSkills: number;
  verifiedCount: number;
  unverifiedCount: number;
  certificationCount: number;
  expiringCertificationCount: number;
  averageSkillsPerEmployee: number;
  skillsByCategory: Record<string, number>;
  skillsByLevel: Record<string, number>;
}

function base(wsId: string, deptId: string, employeeId: string) {
  return `/workspaces/${wsId}/departments/${deptId}/employees/${employeeId}/skills`;
}

export const employeeSkillService = {
  list: (wsId: string, deptId: string, employeeId: string) =>
    apiClient.get<PageResponse<EmployeeSkillResponse>>(`${base(wsId, deptId, employeeId)}`),

  getById: (wsId: string, deptId: string, employeeId: string, skillId: string) =>
    apiClient.get<EmployeeSkillResponse>(`${base(wsId, deptId, employeeId)}/${skillId}`),

  create: (wsId: string, deptId: string, employeeId: string, data: CreateEmployeeSkillRequest) =>
    apiClient.post<EmployeeSkillResponse>(`${base(wsId, deptId, employeeId)}`, data),

  update: (wsId: string, deptId: string, employeeId: string, skillId: string, data: UpdateEmployeeSkillRequest) =>
    apiClient.put<EmployeeSkillResponse>(`${base(wsId, deptId, employeeId)}/${skillId}`, data),

  delete: (wsId: string, deptId: string, employeeId: string, skillId: string) =>
    apiClient.delete<void>(`${base(wsId, deptId, employeeId)}/${skillId}`),

  verify: (wsId: string, deptId: string, employeeId: string, skillId: string) =>
    apiClient.put<EmployeeSkillResponse>(`${base(wsId, deptId, employeeId)}/${skillId}/verify`),

  unverify: (wsId: string, deptId: string, employeeId: string, skillId: string) =>
    apiClient.delete<EmployeeSkillResponse>(`${base(wsId, deptId, employeeId)}/${skillId}/verify`),

  getStats: (wsId: string, deptId: string) =>
    apiClient.get<EmployeeSkillStatistics>(`/workspaces/${wsId}/departments/${deptId}/skills/stats`),
};
